### PR TITLE
Changes for sync trigger, keys, lease management.

### PR DESCRIPTION
**Pull request by AmeyaRele**
### Issue describing the changes in this PR

Changes for functions host to be able to support sync trigger in Managed App Environment.

- [IsSyncTriggersEnvironment()](https://github.com/Azure/azure-functions-host/blob/dev/src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs#L192-L225) check modified to pass.
- Method [BuildSetTriggersRequest](https://github.com/Azure/azure-functions-host/blob/dev/src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs#L662-L671) modified to read endpoint from FUNCTIONS_API_SERVER.
- Add [request headers](https://github.com/Azure/azure-functions-host/blob/268674893728a3b7d5a197edc5b961cd5e095849/src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs#L714-L719) for Managed App Environment.
- Update host id for hash blob to use WEBSITE_SITE_NAME, modify this [check](https://github.com/Azure/azure-functions-host/blob/7b828af7614660bc4b902f5e931517a5352f0968/src/WebJobs.Script/Host/ScriptHostIdProvider.cs#L55) to pass.
- [WebsiteAuthEncryptionKey](https://github.com/Azure/azure-functions-host/blob/54941dba8cece0c93744c9bc6a23ea469d0873b7/src/WebJobs.Script.WebHost/Security/SimpleWebTokenHelper.cs#L93) - Managed App environment wants to leverage WEBSITE_AUTH_ENCRYPTION_KEY.
- [KubernetesDistributedLockManager](https://github.com/Azure/azure-functions-host/blob/54941dba8cece0c93744c9bc6a23ea469d0873b7/src/WebJobs.Script/ScriptHostBuilderExtensions.cs#L301) - Requires this to fail since managed app environment is using storage account for secrets.
- [KubernetesEventGenerator](https://github.com/Azure/azure-functions-host/blob/268674893728a3b7d5a197edc5b961cd5e095849/src/WebJobs.Script.WebHost/WebHostServiceCollectionExtensions.cs#L117-L119) - Want this to be successful for trace logs to be published in functions logs.
- [EnvironmentReadyCheckMiddleware](https://github.com/Azure/azure-functions-host/blob/268674893728a3b7d5a197edc5b961cd5e095849/src/WebJobs.Script.WebHost/WebJobsApplicationBuilderExtension.cs#L41-L43) - Want this to be successful in managed app environment.

### Pull request checklist

* [ ] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [ ] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [ ] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [ ] My changes **do not** require diagnostic events changes
    * Otherwise: I have added/updated all related diagnostic events and their documentation (Documentation issue linked to PR)
* [x] I have added all required tests (Unit tests, E2E tests)

### CODE REVIEW

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs [hunk 1]
```diff
@@ -659,9 +659,9 @@ private string GetDefaultDurableV2HubName()
         internal HttpRequestMessage BuildSetTriggersRequest()
         {
             var url = default(string);
-            if (_environment.IsKubernetesManagedHosting())
+            if (_environment.IsKubernetesManagedHosting() || _environment.IsManagedAppEnvironment())
             {
-                var buildServiceHostname =
+                var buildServiceHostname = _environment.GetEnvironmentVariable("FUNCTIONS_API_SERVER") ??
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: I'd also generalize this var name to just "hostName" now as well

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs [hunk 2]
```diff
@@ -659,9 +659,9 @@ private string GetDefaultDurableV2HubName()
         internal HttpRequestMessage BuildSetTriggersRequest()
         {
             var url = default(string);
-            if (_environment.IsKubernetesManagedHosting())
+            if (_environment.IsKubernetesManagedHosting() || _environment.IsManagedAppEnvironment())
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: In the API server PR, there was discussion on https support, and I believe it was added. Should you use https here?

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs [hunk 3]
```diff
@@ -711,6 +711,13 @@ private async Task<(bool Success, string ErrorMessage)> SetTriggersAsync(string
                 request.Headers.Add(ScriptConstants.SiteTokenHeaderName, token);
                 request.Content = new StringContent(content, Encoding.UTF8, "application/json");
 
+                if (_environment.IsManagedAppEnvironment())
+                {
+                    request.Headers.Add("K8SE-APP-NAME", _environment.GetEnvironmentVariable("CONTAINER_APP_NAME"));
+                    request.Headers.Add("K8SE-APP-NAMESPACE", _environment.GetEnvironmentVariable("CONTAINER_APP_NAMESPACE"));
+                    request.Headers.Add("K8SE-APP-REVISION", _environment.GetEnvironmentVariable("CONTAINER_APP_REVISION"));
+                }
+
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: I think the following if clause should now be an else?

File: src/WebJobs.Script/Environment/EnvironmentExtensions.cs [hunk 4]
```diff
@@ -254,6 +254,16 @@ public static bool IsWindowsAzureManagedHosting(this IEnvironment environment)
             return environment.IsAppService() && RuntimeInformation.IsOSPlatform(OSPlatform.Windows);
         }
 
+        /// <summary>
+        /// Gets a value indicating whether the application is running in Centauri environment.
```
Comment (mathewc) [inline src/WebJobs.Script/Environment/EnvironmentExtensions.cs]: Let's avoid using Centauri project name, and use "Managed App Environment" instead

File: test/WebJobs.Script.Tests/FunctionsSyncServiceTests.cs [hunk 5]
```diff
@@ -109,24 +109,25 @@ public async Task UnhandledExceptions_HandledInCallback()
...
+        [InlineData(false, false, true, false, false, true, false, false)] // no encryption key
         // note: normally linux dedicated would have AzureWebsiteInstanceId set.
         // However the test will always catch it as Windows because it checks the OS of the running process.
         // Setting AzureWebsiteInstanceId to null will make it fail the IsAppServiceWindows, and it shouldn't affect the check for dedicated linux
-        [InlineData(false, true, false, true, false, false, false)] // dedicated linux
-        [InlineData(false, true, false, false, false, false, true)] // dedicated linux
-        public void IsSyncTriggersEnvironment_StandbyMode_ReturnsExpectedResult(bool isAppService, bool hasEncryptionKey, bool isConsumptionLinuxOnAtlas, bool isConsumptionLinuxOnLegion, bool standbyMode, bool containerReady, bool expected)
+        [InlineData(false, true, false, true, false, false, false, false)] // dedicated linux
+        [InlineData(false, true, false, false, false, false, false, true)] // dedicated linux
+        [InlineData(false, true, false, false, false, false, true, true)] // Centauri
```
Comment (mathewc) [inline test/WebJobs.Script.Tests/FunctionsSyncServiceTests.cs]: Remove Centauri here and elsewhere

File: src/WebJobs.Script.WebHost/WebJobsApplicationBuilderExtension.cs [hunk 6]
```diff
@@ -38,7 +38,7 @@ public static IApplicationBuilder UseWebJobsScriptHost(this IApplicationBuilder
             builder.UseMiddleware<HttpRequestBodySizeMiddleware>();
             builder.UseMiddleware<SystemTraceMiddleware>();
             builder.UseMiddleware<HostnameFixupMiddleware>();
-            if (environment.IsAnyLinuxConsumption() || environment.IsKubernetesManagedHosting())
+            if (environment.IsAnyLinuxConsumption() || environment.IsKubernetesManagedHosting() || environment.IsManagedAppEnvironment())
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/WebJobsApplicationBuilderExtension.cs]: Similar to how we introduced IsAnyLinuxConsumption, let's add an IsAnyKubernetesEnvironment check to simplify the currently 2 places where we're doing the check for Kubernetes

File: src/WebJobs.Script/Environment/EnvironmentExtensions.cs [hunk 7]
```diff
@@ -254,6 +254,16 @@ public static bool IsWindowsAzureManagedHosting(this IEnvironment environment)
...
+        /// </summary>
+        /// <param name="environment">The environment to verify.</param>
+        /// <returns><see cref="true"/> if running in Managed App environment; otherwise, false.</returns>
+        public static bool IsManagedAppEnvironment(this IEnvironment environment)
```
Comment (mathewc) [inline src/WebJobs.Script/Environment/EnvironmentExtensions.cs:272]: Add unit tests to EnvironmentExtensionsTests for any new logic you're introducing

File: src/WebJobs.Script/Environment/EnvironmentExtensions.cs [hunk 8]
```diff
@@ -320,7 +330,8 @@ public static bool IsAppService(this IEnvironment environment)
         public static bool IsKubernetesManagedHosting(this IEnvironment environment)
         {
             return !string.IsNullOrEmpty(environment.GetEnvironmentVariable(KubernetesServiceHost))
-            && !string.IsNullOrEmpty(environment.GetEnvironmentVariable(PodNamespace));
+            && !string.IsNullOrEmpty(environment.GetEnvironmentVariable(PodNamespace))
```
Comment (mathewc) [inline src/WebJobs.Script/Environment/EnvironmentExtensions.cs:343]: Update the existing IsKubernetesManagedHosting_ReturnsExpectedResult to test your new logic as well

File: src/WebJobs.Script/Environment/EnvironmentExtensions.cs [hunk 9]
```diff
@@ -262,7 +272,7 @@ public static bool IsWindowsAzureManagedHosting(this IEnvironment environment)
         /// <returns><see cref="true"/> if running in a Linux Consumption App Service app; otherwise, false.</returns>
         public static bool IsAnyLinuxConsumption(this IEnvironment environment)
         {
-            return environment.IsLinuxConsumptionOnAtlas() || environment.IsLinuxConsumptionOnLegion();
+            return (environment.IsLinuxConsumptionOnAtlas() || environment.IsLinuxConsumptionOnLegion()) && !environment.IsManagedAppEnvironment();
```
Comment (mathewc) [inline src/WebJobs.Script/Environment/EnvironmentExtensions.cs:285]: There isn't an existing test for this method, but please add one that tests the existing logic as well as your new logic

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs (hunk 3 above)
Comment (AmeyaRele) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: Changed the clause to 'else if'

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs (hunk 2 above)
Comment (krishnajaju) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: Env var `FUNCTIONS_API_SERVER` is already set with the https protocol used. Env var looks like this 
`FUNCTIONS_API_SERVER=https://<appname>.internal.<DNSsuffix>`.

**Review by mathewc — APPROVED** [review]

**Review by krishnajaju — APPROVED** [review]

**Comment by mathewc** [comment]
The PR description mentions "keys, lease management" but there is none of that in this PR right? recommend updating the PR description to reflect that it only contains sync triggers changes?

**Comment by AmeyaRele** [comment]
> The PR description mentions "keys, lease management" but there is none of that in this PR right? recommend updating the PR description to reflect that it only contains sync triggers changes?

There are a few changes for keys and lease management. The check IsKubernetesManagedHosting() was modified for those changes. 
[WebsiteAuthEncryptionKey](https://github.com/Azure/azure-functions-host/blob/54941dba8cece0c93744c9bc6a23ea469d0873b7/src/WebJobs.Script.WebHost/Security/SimpleWebTokenHelper.cs#L93) - Managed App environment wants to leverage WEBSITE_AUTH_ENCRYPTION_KEY.
[KubernetesDistributedLockManager](https://github.com/Azure/azure-functions-host/blob/54941dba8cece0c93744c9bc6a23ea469d0873b7/src/WebJobs.Script/ScriptHostBuilderExtensions.cs#L301) - Requires this to fail since managed app environment is using storage account for secrets.
[KubernetesEventGenerator](https://github.com/Azure/azure-functions-host/blob/268674893728a3b7d5a197edc5b961cd5e095849/src/WebJobs.Script.WebHost/WebHostServiceCollectionExtensions.cs#L117-L119) - Want this to be successful for trace logs to be published in functions logs.
[EnvironmentReadyCheckMiddleware](https://github.com/Azure/azure-functions-host/blob/268674893728a3b7d5a197edc5b961cd5e095849/src/WebJobs.Script.WebHost/WebJobsApplicationBuilderExtension.cs#L41-L43) - Want this to be successful in managed app environment.

**Pull request merged**